### PR TITLE
workspace: Set default value for LocalPathsOrder in query from DB for recent workspaces

### LINKED ISSUE #16119: Open recent folders list keeps showing empty items for removed folders
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

When opening the "Open Recent..." list from `File > Open Recent...`  or from command palette, the list shows some empty items at the bottom of list. When I click any of these empty items, random folder opens, not able to find a pattern here.

### Environment

Zed: v0.147.2 (Zed)
OS: macOS 14.5.0
Memory: 18 GiB
Architecture: aarch64

### If applicable, add mockups / screenshots to help explain present your vision of the feature

https://github.com/user-attachments/assets/3e696eea-7d9f-4bab-9328-9064f45109f3

### If applicable, attach your Zed.log file to this issue.
<details><summary>zed.log</summary><pre>
2024-08-11T23:56:06.661569Z [INFO] Language server with id 2 sent unhandled notification eslint/noLibrary:
{
  "source": {
    "uri": "file:///Users/harmeetsingh/Documents/code/playwright/app/ToDoApp.ts"
  }
}
2024-08-11T23:56:06.662264Z [INFO] Language server with id 2 sent unhandled notification eslint/noLibrary:
{
  "source": {
    "uri": "file:///Users/harmeetsingh/Documents/code/playwright/Components/TaskList.ts"
  }
}
2024-08-11T23:56:06.664544Z [INFO] Language server with id 2 sent unhandled notification eslint/noLibrary:
{
  "source": {
    "uri": "file:///Users/harmeetsingh/Documents/code/playwright/tests-examples/new/new-tests.app.spec.ts"
  }
}
2024-08-11T23:56:08.19885Z [INFO] downloading language server "package-version-server"
2024-08-11T23:56:08.199983Z [INFO] starting language server. binary path: "/Users/harmeetsingh/Library/Application Support/Zed/languages/package-version-server/package-version-server-v0.0.4", working directory: "/Users/harmeetsingh/Documents/code/playwright", args: []
2024-08-11T23:56:08.2575Z [INFO] downloading language server "json-language-server"
2024-08-11T23:56:08.258648Z [INFO] Node runtime install_if_needed
2024-08-11T23:56:08.344073Z [INFO] starting language server. binary path: "/Users/harmeetsingh/Library/Application Support/Zed/node/node-v22.5.1-darwin-arm64/bin/node", working directory: "/Users/harmeetsingh/Documents/code/playwright", args: ["/Users/harmeetsingh/Library/Application Support/Zed/languages/json-language-server/node_modules/vscode-langservers-extracted/bin/vscode-json-language-server", "--stdio"]
2024-08-12T07:37:18.336272Z [WARN] request completed with error: request or operation took longer than the configured timeout time
2024-08-12T07:37:18.34356Z [ERROR] auto-update failed: error:request or operation took longer than the configured timeout time

Caused by:
    [28] Timeout was reached
2024-08-12T15:27:34.798098Z [INFO] building git repository, `.git` path in the worktree: ".git"
2024-08-12T15:27:34.825725Z [INFO] Initializing default prettier with plugins {}
2024-08-12T15:27:34.826184Z [INFO] starting language servers for TypeScript: vtsls, eslint, graphql
2024-08-12T15:27:34.826244Z [INFO] starting language servers for JSON: json-language-server, package-version-server
2024-08-12T15:27:34.826296Z [INFO] starting language servers for TypeScript: vtsls, eslint, graphql
2024-08-12T15:27:34.826347Z [INFO] starting language servers for TypeScript: vtsls, eslint, graphql
2024-08-12T15:27:34.826402Z [INFO] starting language servers for TypeScript: vtsls, eslint, graphql
2024-08-12T15:27:34.860056Z [WARN] request completed with error: failed to connect to the server
2024-08-12T15:27:35.233395Z [INFO] Language server with id 2 sent unhandled notification eslint/noLibrary:
{
  "source": {
    "uri": "file:///Users/harmeetsingh/Documents/code/playwright/tests-examples/new/new-tests.app.spec.ts"
  }
}
2024-08-12T15:27:35.233639Z [INFO] Language server with id 2 sent unhandled notification eslint/noLibrary:
{
  "source": {
    "uri": "file:///Users/harmeetsingh/Documents/code/playwright/components/Input.ts"
  }
}
2024-08-12T15:27:35.233697Z [INFO] Language server with id 2 sent unhandled notification eslint/noLibrary:
{
  "source": {
    "uri": "file:///Users/harmeetsingh/Documents/code/playwright/Components/TaskList.ts"
  }
}
2024-08-12T15:27:35.233749Z [INFO] Language server with id 2 sent unhandled notification eslint/noLibrary:
{
  "source": {
    "uri": "file:///Users/harmeetsingh/Documents/code/playwright/app/ToDoApp.ts"
  }
}
</pre></details>

### PR DESCRIPTION
**Pull request by CharlesChen0823**
Closes #16119 

I found the value `LocalPathsOrder` in `self.recent_workspaces()` not used before #12844, after #12844, using `LocalPathsOrder` to display recent project, but the value in DB might be empty for older release.  So I add default value for `LocalPathsOrder` in query from DB.

Don't known if this is expect?

Release Notes:

- N/A

### CODE REVIEW

**Comment by SomeoneToIgnore** [comment]
сс @eth0net in case you want to check on that

File: crates/workspace/src/persistence.rs [hunk 1]
```diff
@@ -767,6 +767,12 @@ impl WorkspaceDb {
         let dev_server_projects = self.dev_server_projects()?;
 
         for (id, location, order, dev_server_project_id) in self.recent_workspaces()? {
+            let temp_order = order.clone();
```
Comment (SomeoneToIgnore) [inline crates/workspace/src/persistence.rs:770]: 1. `.clone()` is not needed here
2. This seems to be the wrong place to fix things in, but maybe I misunderstand the root cause.
If this is the right fix, I think it belongs to https://github.com/zed-industries/zed/blob/1e39d407c24dbe5a0eeee3f12b2138129bcb95fa/crates/workspace/src/persistence/model.rs#L99

I understand it's not possible to fix it directly there though (ergo hoped for @eth0net to chime in with better ideas), at least it's obvious that there's more than one `SELECT .. local_paths_order, ..` around — seems that we have to fix them all too?

**Review by SomeoneToIgnore — COMMENTED** [review]
(I remember about https://github.com/zed-industries/zed/pull/15396 and have to review it soon, hopefully, sorry for delays)

I think you've spotted something near the root cause in this PR, which is nice.
This PR would very benefit from
* a reproduction steps description — I do not think it's possible to write a test for this? But it's much nicer when the reviewer can also verify the code manually at least.

* a brief explanation on what are we doing — I am totally confused why, having a full control over storing and reading from the database, we do 

```Rust
// get some data from the database with SQL query
for (id, location, order, dev_server_project_id) in self.recent_workspaces()? {
    // if some of this data is empty
    let order = if order.order().len() == 0 {
        // use another part of this data to stub another piece of the data from the same SQL query
        LocalPathsOrder::default_for_paths(&location)
    } else {
        // or use the data from the database without any substitutions by another data from the database
        order
    };
//.............
}
```

This does not make any sense — if the input data is wrong, we can sanitize it before INSERT/UPDATE.

**Comment by SomeoneToIgnore** [comment]
I believe that we need to fix blank entries shown in the list, but do that the correct way.

**Comment by SomeoneToIgnore** [comment]
Reading the last comment a few times, I start to understand that the issue is with the fact that old code had spoiled the DB and now the new code reads it incorrectly?

That changes things indeed, we are not doing any data migrations right now IIRC.

**Comment by CharlesChen0823** [comment]
yes, the old code serialized data into DB don't have this column data, when the new code query result is empty.

**Comment by SomeoneToIgnore** [comment]
Thank you for the explanations.
It would save a lot of time on the decision-making if was put as a description in the PR.

In this particular case, let's close the issue indeed: I'll think of an SQL command to clean the odd database state and post into that issue.
The problematic code existed in a few versions for a limited amount of people, is not very widespread — and clearing recent projects list have to be a build-in functionality anyway.

**Pull request closed without merging**